feat : 커플 연결시 채팅방 연결

asklas opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요. 이미지 or 움짤을 통해 확인 가능하도록 해주세요 -->
- 커플 연결시 자동으로 채팅방도 연결되도록 설정하였습니다
-
-

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- member Entity에 채팅방 ID 추가하였습니다
- 파트너가 업데이트 될 때 상대도 동기화 되도록 하였습니다
-

## ✋ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 파트너 등록 메서드를 둘로 나누어 요청/수락 하도록 할 것인지, 이미 상호 동의되었다는 가정 하에 한번에 등록할 것인지에 대해 의견 주시면 감사하겠습니다.

Commit:
feat: 커플 연결시 채팅방 연결

seosangwon commented:
요청/수락이 나을 것 같습니다!

youjungHwang commented on src/main/java/org/example/flowday/domain/member/service/MemberService.java:
```diff
@@ -361,9 +375,12 @@ public void updatePartnerId(Long id, MemberDTO.UpdatePartnerIdRequestDTO dto) {
         if (dto.getPartnerId() != null) {
             member.setPartnerId(dto.getPartnerId());
         }
```
혹시 이부분이 어떤 의미일까요?

asklas replied:
파트너 ID를 설정하는 부분입니다. 지금 보니 조건 검사는 필요 없어 보이네요 수정하겠습니다. 😄

asklas replied:
수정 하였습니다

youjungHwang commented on src/main/java/org/example/flowday/domain/member/service/MemberService.java:
```diff
@@ -361,9 +375,12 @@ public void updatePartnerId(Long id, MemberDTO.UpdatePartnerIdRequestDTO dto) {
         if (dto.getPartnerId() != null) {
             member.setPartnerId(dto.getPartnerId());
         }
+        member.setChattingRoomId(chatService.registerChatRoom(LocalDateTime.now()));
```
요청을 보내고 아직 수락하기 전에 채팅방을 만드는 로직인거죠? 해당 코드처럼 갈지 아니면 수락하고 난 후에 채팅방을 만들지 같이 고민해보면 좋을 것같습니다. 😁

asklas replied:
네 맞습니다. 커플 등록이 확정되었다는 가정 하에 작성된 로직이라, 요청/수락 방식을 선택한다면 변경되어야 할 듯 합니다.

youjungHwang commented on src/main/java/org/example/flowday/domain/member/service/MemberService.java:
```diff
@@ -351,7 +351,21 @@ public void updateRelationshipStartDate(Long id, MemberDTO.UpdateRelationshipSta
+        Member member = memberRepository.findById(myId)
+                .orElseThrow(() -> new IllegalArgumentException("Member not found with id: " + myId));
+
+        member.setPartnerId(partnerId);
```
객체 생성에 setter 쓰시는 것 같은데, 혹시 이유가 있으실까요?🤔🤗

asklas replied:
질문의 의도를 파악하기 어렵습니다. 혹시 생각하시는 대안의 예시나, 명확한 질문의 의도를 알 수 있을까요?

youjungHwang replied:
네네, 예를들어서 빌더패턴이나 정적 팩토리 메서드를 사용하는 방법등이 있는데 그중에서 setter방식을 쓰시는 것 같아서요. 혹시 해당 방법의 장점이 어떤게 있을까해서 여쭈어보게되었습니다.

asklas replied:
제 생각에는 조금 오해가 있는 듯 하여 일단 말씀 드리자면, Member 안의 partnerId 필드는 단순한 Long타입 데이터 입니다. 그냥 ID값만 가지고 있고 다른 Member를 참조하지 않습니다.
따라서 단순한 setter로 id값만 입력합니다.

제가 잘못 이해한거라면 말씀해주시면 감사하겠습니다.

youjungHwang replied:
아닙니다. 객체 생성 방식이 여러가지니까 어떤의도로 사용하시는지 궁금했어요 ㅎㅎ 말씀 이해했습니다!

asklas replied:
아 이제 이해했습니다. Setter가 편해서 사용했는데, 리팩토링 할 때 관련 부분 고려해봐야겠네요. 의견 감사합니다 😄

youjungHwang commented on src/main/java/org/example/flowday/domain/member/service/MemberService.java:
```diff
@@ -351,7 +351,21 @@ public void updateRelationshipStartDate(Long id, MemberDTO.UpdateRelationshipSta
+    @Transactional
+    public void acceptAddPartnerId(Long partnerId, Long myId, Long chattingRoomId) {
+
+        Member member = memberRepository.findById(myId)
+                .orElseThrow(() -> new IllegalArgumentException("Member not found with id: " + myId));
```
해당 부분 다른 메서드도 공통적으로 쓰이는 것 같아, private 메서드로 분리하는 방법도 고려해보시면 좋을 것 같습니다.

asklas replied:
SELECT * FROM ... 형식의 쿼리는 리팩토링 과정에서 모두 제거하려 하였는데 타 도메인에서도 많이 사용되는 탓에 일단 남겨두었습니다
existById로 변경한 후 분리하는 방식으로 변경하는게 좋을 것 같네요 의견 감사합니다 👍

youjungHwang replied:
네네 그럼 추가적으로 "SELECT * FROM ... 형식의 쿼리는 리팩토링 과정에서 모두 제거하려 하였는데" -> 이렇게하면, 메서드 안에서 id의 유무를 어떤식으로 파악할 예정이실까요?? 괜찮은 방법이면 저도 바로 도입해보려합니다!!

asklas replied:
id 값이 존재하는지 체크만 하면 되는거라면 existsById를 사용하고, 

값을 가져오기까지 해야 한다면

    @Query("SELECT m.role AS role, m.id AS id, m.pw AS pw FROM Member m WHERE m.loginId = :loginId")
    Optional<Map<String, Object>> findSecurityInfoByLoginId(@Param("loginId") String loginId); // 로그인 ID로 검색

    @Modifying
    @Transactional
    @Query("UPDATE Member m SET m.refreshToken = :refreshToken WHERE m.loginId = :loginId")
    void updateRefreshToken(@Param("loginId") String loginId, @Param("refreshToken") String refreshToken);
    
위의 예시처럼 필요한 값만 가져가서 변경, Update를 사용하여 반영 하도록 할 예정입니다.

변경하는 이유는 JMeter로 부하테스트를 진행했을 때 전체를 가져오는 쿼리보다 일부분을 가져가서 수정하는 쿼리가 반응 속도 면에서 안정적인 응답을 하는 것을 확인하였기 때문입니다.

다른 부분은 모두 리팩토링하여 MemberRepository에 반영되어 있으니, 필요하시다면 develop 브런치에 작성되어 있는 코드를 확인해주시면 감사하겠습니다.

youjungHwang replied:
오.. 네네 해당 부분 확인하고 저도 반영하는거 고려해보겠습니다. 감사합니다!!

youjungHwang reviewed (commented):
수고하셨습니다. 그리고 `@Data` 보단 필요한 부분만 어노테이션 쓰는 방법도 있을 것 같습니다!
저도 요청, 수락이 나을 것 같습니다!

kang-ye-jin reviewed (approved):
요청, 수락이 좋을 것 같습니다~ 고생하셨습니다!

Commit:
refactor:
- id값으로 member를 찾는 코드를 분리하였습니다
- 일부 쿼리문 최적화 하였습니다
fix:
- id/pw 찾기 기능이 인증을 요구하였던 버그를 수정하였습니다
